### ISSUE TITLE
 Unable to launch game in either windowed or fullscreen mode

### ISSUE THREAD

**Issue by clin1234** [freeorion/freeorion#4978]
Bug Report
==========

Unable to launch game in either windowed or fullscreen mode

Environment
-----------

* **FreeOrion Version**: 2024-07-12.0739705
* **Operating System**: Windows 11 Pro
* **Graphic card used**: Nvidia Mobile RTX 3060
* **Fetched as**
  * Binary release
  * Weekly development build

Description
-----------

config.xml:
```
<?xml version="1.0"?>
<XMLDoc>
  <video>
    <fullscreen>
      <reset>0</reset>
      <width>1463</width>
      <height>823</height>
    </fullscreen>
    <fullscreen/>
  </video>
  <version>
    <string>v0.5+ [build 2024-07-12.0739705] MSVC 2022 Win32</string>
  </version>
  <video/>
</XMLDoc>
```

[freeorion.log](https://github.com/user-attachments/files/16257615/freeorion.log)

Expected Result
---------------
Game to run

Steps to reproduce
------------------

* Click on either windowed or fullscreen version of test build

**Comment by Blue-Award** [comment]
Have you run the installer, or just unzipped files? 

I usually run the installer but this time around I tried unzipping new version from "binary only" release. It did not indeed work for me, either (Windows 10). Cmd line window launches and then the game closes.

Meanwhile, after having done full reinstallation from the installer release - it runs.

For some reason the exectuables in "binary only" release are different sizes to the one that comes with full installation. Dunno why the difference

**Comment by Blue-Award** [comment]
although in your case log shows something about unitialized python, in my case there's no log even, so probably different issue

**Comment by clin1234** [comment]
> Have you run the installer, or just unzipped files?
> 
> I usually run the installer but this time around I tried unzipping new version from "binary only" release. It did not indeed work for me, either (Windows 10). Cmd line window launches and then the game closes.
> 
> Meanwhile, after having done full reinstallation from the installer release - it runs.
> 
> For some reason the exectuables in "binary only" release are different sizes to the one that comes with full installation. Dunno why the difference

The installer: source is [this](https://sourceforge.net/projects/freeorion/files/FreeOrion/Test/FreeOrion_2024-07-12.0739705_Test_Win32_Setup.exe/download)

**Comment by geoffthemedio** [comment]
My guess is that there's some problem with the path to Python given the non-Latin characters in the username in this path:
`19:55:16.710936 {0x00005d0c} [debug] client : i18n.cpp:115 : GetDefaultStringTableFileName returning: C:/Users/ךינשגכהד/scoop/apps/freeorion-test/current/default/stringtables/en.txt`

which leads to Python not being found and:
````
19:55:16.849973 {0x0000506c} [debug] client : PythonCommon.cpp:17 : Initializing FreeOrion Python interface
19:55:16.851024 {0x0000506c} [error] client : PythonCommon.cpp:49 : Unable to initialize Python interpreter
19:55:16.851024 {0x0000506c} [error] client : PythonParser.cpp:63 : Python parse given non-initialized python!
````

@clin1234 Can you try installing (or moving the FreeOrion root directory) somewhere with just Latin characters in the path?

@o01eg Any other ideas?

**Comment by o01eg** [comment]
No, I suppose this is an issue. Not sure how to solve it.

**Comment by geoffthemedio** [comment]
Windows APIs expect UTF-16 strings, including for paths, so I suppose what's needed to to make sure any paths are formatted like that before using it. There are some examples in the code already of doing that, usually within #ifdef blocks checking MSC_VER. Probably difficult to troubleshoot if you're not on Windows though.

**Comment by Blue-Award** [comment]
> Windows APIs expect UTF-16 strings, including for paths, so I suppose what's needed to to make sure any paths are formatted like that before using it. There are some examples in the code already of doing that, usually within #ifdef blocks checking MSC_VER. Probably difficult to troubleshoot if you're not on Windows though.

I have an issue like that trying to load multiplayer saves. I rarely do it, I resorted to creating a symlink to my original %appdata%, and setting %appdata% to that symlink for FO session where I want to load multiplayer saves.

Looks like what you're saying would fix that, too. Do you have an example of code that does the relevant thing handy?

**Comment by geoffthemedio** [comment]
See: https://github.com/freeorion/freeorion/blob/master/GG/src/dialogs/FileDlg.cpp#L353-L370

**Comment by Blue-Award** [comment]
It might be a treasure trove of issues. I tried reproducing this on my Windows 10 machine, and I do get a crash, but in a different place. 

<pre>[error] client : ClientUI.cpp:1079 : Unable to load texture "C:/????????/FreeOrion/default/data/art/icons/buttons/close_mouseover.png"
reason: file_stream_device: failed to open file for reading: iostream stream error
[error] client : chmain.cpp:318 : main() caught exception(std::exception): Attempted to contruct subtexture from invalid texture</pre>

Right before that, the log also has 
<pre>[error] client : StringTable.cpp:149 : StringTable::Load failed to read file at path: C:/ךינשגכהד/FreeOrion/default/stringtables/en.txt</pre>

Meanwhile, for @clin1234 it seems to load the stringtable successfully, curious. Maybe some difference between APIs of Win11 vs Win10?

I'll have a look. I'm mostly interested in fixing the issue I have with multiplayer saves, which is about %appdata% having unicode char rather than FreeOrion's parent path having such char, but fixing just that would feel bit of incomplete without fixing the issue here. 

Somewhat worried how many instances of this are there for the main FO path, but let's see how much I manage to find/fix. Need to see if I am even able to reproduce the thing with Python after fixing the earlier failures I have - I have Python installed separately, maybe it can affect things?

**Comment by clin1234** [comment]
> It might be a treasure trove of issues. I tried reproducing this on my Windows 10 machine, and I do get a crash, but in a different place.
> 
> [error] client : ClientUI.cpp:1079 : Unable to load texture "C:/????????/FreeOrion/default/data/art/icons/buttons/close_mouseover.png"
> reason: file_stream_device: failed to open file for reading: iostream stream error
> [error] client : chmain.cpp:318 : main() caught exception(std::exception): Attempted to contruct subtexture from invalid texture
> Right before that, the log also has
> 
> [error] client : StringTable.cpp:149 : StringTable::Load failed to read file at path: C:/ךינשגכהד/FreeOrion/default/stringtables/en.txt
> Meanwhile, for @clin1234 it seems to load the stringtable successfully, curious. Maybe some difference between APIs of Win11 vs Win10?
> 
> I'll have a look. I'm mostly interested in fixing the issue I have with multiplayer saves, which is about %appdata% having unicode char rather than FreeOrion's parent path having such char, but fixing just that would feel bit of incomplete without fixing the issue here.
> 
> Somewhat worried how many instances of this are there for the main FO path, but let's see how much I manage to find/fix. Need to see if I am even able to reproduce the thing with Python after fixing the earlier failures I have - I have Python installed separately, maybe it can affect things?

If it helps, I have my encoding set to UTF-8

**Comment by Blue-Award** [comment]
@clin1234 can you confirm if running FreeOrion from a different folder, one that contains only ANSI characters, works for you?

**Comment by clin1234** [comment]
Extracting to a folder containing only ASCII characters, and running the game from there works

**Comment by Blue-Award** [comment]
I looked into my win 10 and it has this "BETA" option to turn on treating non-unicode conforming apps as if they were doing things in UTF-8. Not sure since when this option is there. When I turn it on, I can reproduce the issue the way @clin1234 has it, i.e. my game actually loads the stringtables and the textures (and then fails on unitialized python)

This gives me vibe that utf16to8 is what was actually BREAKING my stuff in the first place. @geoffthemedio you said it yourself that win APIs expected UTF16. So why there is conversion to UTF8? boost's path.native() seems to be UTF16. So why convert it to UTF8 which seems actually NOT TO WORK unless Windows is set up specifically to treat non-unicode conforming apps to be UTF8?

I suppose I have to check what happens if I switch back Windows 10 from that beta setting of supporting UTF8 and remove the utf16to8 conversion

**Comment by Blue-Award** [comment]
I might be derailing the original issue here a bit but starting to suspect making FreeOrion properly unicode, at least in windows, may be too big of a task for me. I tried pursuing this texture loading for instance. Like - gil::read_image has a wstring version, so that suggests it should work directly on path.native() without prior utf16to8 conversion. But ultimately it does its own conversion using wcstombs and that produces ??????? for me which is even worse, because with converting to utf8 I can at least force Windows to assume that's what the apps are going to do (seems relatively recent setting for windows, tho)

Seems like there is something missing for FreeOrion to make wcstombs work properly in this scenario?

LIke it really seems there should be no utf16toutf8 conversion necessary. Basically as is, FreeOrion seems just not to be a Unicode app (as far as Windows is concerned, anyway). I think I've read some blog post about unicode support back in the day, and also some of my prior work experience, suggests that (at least windows) apps should pretty much use wstring everywhere rather than string to be "properly" unicode? And likely do some other stuff with locale etc. Bit hazy on this.

I'm glad at least that I've learned that nowadays you can force WIndows to treat non-unicode apps as trying their best to be unicode-adjacent with UTF8 output (dunno is that default mode for Linux?)


Bottom line tho is that with the UTF8 setting I can at least exactly reproduce the original issue. And my own issue with %appdata% prolly still can use a utf16toutf8 fix that would make it work under this UTF8 windows treatment. We'll see

**Comment by Blue-Award** [comment]
Back to the issue at hand, regarding Python init. 

This looks to come from... DebugLogger failing to convert unicode string... specifically for 
<code>DebugLogger() << "Python program name set to " << Py_GetProgramFullPath();</code>

This is generally worrisome to me, i.e. DebugLogger failling to support wide char array with actual Unicode in it?

Separately, I take issue how this Python init is doing its thing. Seems to me going through Py_DecodeLocale(GetPythonHome().string().c_str(), nullptr) is not right (on Windows) - the unicode characters end up as ??????

Same thing with <code>m_program_name = Py_DecodeLocale((GetPythonHome() / "Python").string().c_str(), nullptr);</code>

 Not like it seems to matter, because Py_GetProgramFullPath(); still just returns full path to FreeOrion.exe with the original unicode characters (and DebugLogger is not happy about them). Curiously in newer Python Py_GetProgramFullPath(); is not supposed to be called before python is fully initialized! (it returns NULL in such case, since Python 3.10)

Either way, whatever that code is trying to do, seems to me that since paths on Windows are already unicode, those should be just copied to m_home_dir and m_program_name rather than going through Py_DecodeLocale. But again, this seems like a separate issue, because even with all that which looks not fine to me, things work, if I only comment out DebugLogger() calls

(By "work" I just mean the application starts - I can't play because AI clients seem crashing, but I don't want to muddle things further as there is plenty of things not working without UTF8 enforcement on win10)

I need to go over those two peels separately - one is why DebugLogger is not happy with unicode wide char array? Any ideas?

Second thing is if the python init is even working properly in unicode path context, ignoring the issue with DebugLogger

**Comment by Blue-Award** [comment]
the DebugLogger issue I suppose can be solved through doing utf16to8 just for debug purposes to successfully print out the thing. Can't use python EncodeLocale cause it's not supposed to be called before Py_Initialize. But  I would still like to understand why the DebugLogger is unhappy with unicode wide char array

**Comment by Blue-Award** [comment]
note DebugLogger also fails here, if I comment out the earlier case:

<code>DebugLogger() << "Python module search path: " << Py_GetPath();</code>

but since it is past Py_Initialize, things progress somewhat further without commenting this one. But it's another place to fix the logging

**Comment by Blue-Award** [comment]
Oh and when I force windows to talk UTF8, then those Py_DecodeLocale actually work (pretty much roundtripping the unicode path, without ending up with ????? chars), and then... DebugLogger fails even on <code>DebugLogger() << "Python home set to " << Py_GetPythonHome();</code>

**Comment by geoffthemedio** [comment]
> why DebugLogger is not happy with unicode wide char array? Any ideas?

I've always needed to manually convert to UTF-8 for it to correctly log any UTF-16 strings. There are bits of code that look like they could be intended to auto-convert streamed in text to the output format, but it's never worked for me, and I can't actually find it doing a UTF-16 to UTF-8 conversion, even though it seems to have overloads for wchar_t*.

> This is generally worrisome to me, i.e. DebugLogger failling to support wide char array with actual Unicode in it?

It doesn't really need to, I suppose... It's not using any pathes in the passed in log text to do anything with the Windows API, and that's the only situation in which UTF16 is supplied to the program or needs to be supplied back.

>... with the UTF8 setting I can at least exactly reproduce the original issue. And my own issue with %appdata% prolly still can use a utf16toutf8 fix that would make it work under this UTF8 windows treatment.

Expecting users to change some setting in Windows isn't reasonable.

**Comment by o01eg** [comment]
When I tried to run FreeOrion on Linux from incorrect utf8 path, I got errors with `PythonBase::SetCurrentDir` and `PythonBase::AddToSysPath`. I suppose `GetPythonDir` here returns incorrect std::string too.

```
  File "<string>", line 2
    sys.path.append(r'/tmp/test/999��������999/share/freeorion/default/python')
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
SyntaxError: (unicode error) 'utf-8' codec can't decode byte 0xf4 in position 13: invalid continuation byte
```

**Comment by geoffthemedio** [comment]
> When I tried to run FreeOrion on Linux from incorrect utf8 path, I got errors 

It would probably be more useful to know what happens if you use a valid / existing path with non-Latin characters.

**Comment by Blue-Award** [comment]
I have a fix  both to DebugLogger thing and the somewhat separate wide char path init issue on windows, but please bear with me before I set a PR (should be later today)

**Comment by o01eg** [comment]
> It would probably be more useful to know what happens if you use a valid / existing path with non-Latin characters.

On Linux it works without issues

**Comment by o01eg** [comment]
After reading documentation it looks like `sys.path` allows only correct strings, `PySys_SetPath` is declared deprecated from 3.11 in favor https://docs.python.org/3.12/c-api/init_config.html#c.PyConfig.module_search_paths which accepts `PyWideStringList`

**Comment by Blue-Award** [comment]
I've put up a fix to DebugLogger issue, which includes a fix not to roundtrip wide char path on Windows when trying to Py_SetPythonHome and Py_SetProgramName

But I think there is another issue here, that doc suggests one should do [Py_PreInitialize](https://docs.python.org/3.9/c-api/init_config.html?highlight=preinit#c.Py_PreInitialize) before calling things like Py_DecodeLocale. And seems to me it's not being done?

I mean things look to be working without the Py_PreInitialize, but I would think "better safe than sorry?"
